### PR TITLE
Q-AJN-5: Missing Input Validation

### PR DESCRIPTION
**Pull request by MikeHathaway**
# Description of change
## High level
* Add additional external input validations
  * Update `voteExtraordinary`, `fundingVote`, `screeningVote` to check that each successive vote being cast is using non 0 voting power.
  * Update `_validateSlate` to check that a slate contains a proposal, and that each proposal has >0 funding votes received: https://github.com/code-423n4/2023-05-ajna-findings/issues/465
  * Also fixes a bug where a proposal could be executed even if it received no funding votes by changing a check to `<=`
  * Update invariant tests for the new assertions.
  * Add validation to calldata length as suggested in: https://github.com/code-423n4/2023-05-ajna-findings/issues/280

https://github.com/code-423n4/2023-05-ajna-findings/issues/465

### CODE REVIEW

**Review by prat-gpt — APPROVED** [review]
LGTM

File: src/grants/base/StandardFunding.sol [hunk 1]
```diff
@@ -427,6 +427,9 @@ abstract contract StandardFunding is Funding, IStandardFunding {
         // check that the slate has no duplicates
         if (_hasDuplicates(proposalIds_)) revert InvalidProposalSlate();
 
+        // check that there is a proposal in the slate
```
Comment (grandizzy) [inline src/grants/base/StandardFunding.sol]: suggest better move this at line 428 in same if clause and before checking if it has duplicates

File: src/grants/base/StandardFunding.sol [hunk 2]
```diff
@@ -555,6 +559,9 @@ abstract contract StandardFunding is Funding, IStandardFunding {
             // check that the proposal being voted on is in the top ten screened proposals
             if (_findProposalIndex(voteParams_[i].proposalId, _topTenProposals[currentDistributionId]) == -1) revert InvalidVote();
 
+            // check that the voter isn't attempting to cast a vote with 0 power
+            if (voteParams_[i].votesUsed == 0) revert InvalidVote();
```
Comment (grandizzy) [inline src/grants/base/StandardFunding.sol:586]: Same as above, better check before findProposalIndex

**Review by grandizzy — CHANGES_REQUESTED** [review]

File: src/grants/base/StandardFunding.sol [hunk 3]
```diff
@@ -438,7 +441,8 @@ abstract contract StandardFunding is Funding, IStandardFunding {
             if (_findProposalIndex(proposalIds_[i], _topTenProposals[distributionId_]) == -1) revert InvalidProposalSlate();
 
             // account for fundingVotesReceived possibly being negative
-            if (proposal.fundingVotesReceived < 0) revert InvalidProposalSlate();
+            // block proposals that recieve no positive funding votes from entering a finalized slate
```
Comment (grandizzy) [inline src/grants/base/StandardFunding.sol:457]: Missed that one, also better to do the check in same line 441 before even searching the proposal index

**Review by ith-harvey — COMMENTED** [review]
They suggested that `description_` is checked for non-zero, didn't see that included in PR.

File: src/grants/base/StandardFunding.sol (hunk 1 above)
Comment (MikeHathaway) [reply src/grants/base/StandardFunding.sol]: Updated accordingly

File: src/grants/base/StandardFunding.sol (hunk 2 above)
Comment (MikeHathaway) [reply src/grants/base/StandardFunding.sol:586]: updated

File: src/grants/base/StandardFunding.sol (hunk 3 above)
Comment (MikeHathaway) [reply src/grants/base/StandardFunding.sol:457]: updated

**Comment by MikeHathaway** [comment]
> They suggested that `description_` is checked for non-zero, didn't see that included in PR.

Added

**Review by grandizzy — APPROVED** [review]

**Pull request merged**